### ISSUE TITLE
Scripts using `#!/usr/bin/env python` fail if only `python3` is available`

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#1001]
Some of the examples scripts and the Jinja2 build script (lib/process.py) have `#!/usr/bin/env python` but this fails if only Python 3 is available. (e.g. a clean install of Ubuntu under WSL2 has this problem).

A simple workaround is to do `alias python=python3` but we need to find a robust solution.

**Comment by arporter** [comment]
Ubuntu 20.0.4 has dropped Python 2 altogether and thus only ships with `python3`.

**Comment by arporter** [comment]
Googling around, I found the relevant pep which essentially says, "yes, it's a problem and it's your problem...": https://www.python.org/dev/peps/pep-0394/#for-python-script-publishers

We should really move away from having (executable) python scripts in the examples - they are really a hangover from the old way of doing things and should be replaced with a suitable invocation of `psyclone`.

Not sure what to do about `process.py`. The aforementioned pip talks about re-writing shebangs in installed scripts as required - we could do that.

Alternatively, we could adopt the convention of appending a `3` to Python-3-specific scripts and maybe having these alongside the existing scripts. e.g. psyclone3 and genkernelstub3.

**Comment by hiker** [comment]
One option could be to make process.py a shell script, which checks for various known python names, and uses this to execute process.py?

**Comment by arporter** [comment]
Oh yes - that's a good idea @hiker. In fact, I see that our `psyclone` script also has the same shebang and yet that seemed to work for people yesterday. It's very confusing!

**Comment by hiker** [comment]
I nearly wrote the same, but that's actually not true: the installation process created a script in .local/bin (or wherever), and IT has the right python version (guess they  query the system). This script then starts our script using this python version.

**Comment by hiker** [comment]
I'll take this (we somehow are all assigned ... so I thought I'll let you know :grin: )

**Comment by rupertford** [comment]
python-is-python3 seemed to fix this for me, or at least something similar :-)

**Comment by sergisiso** [comment]
> python-is-python3 seemed to fix this for me, or at least something similar :-)

This is what I use in my WSL

**Comment by rupertford** [comment]
I should probably point out that @sergisiso told me about it in the first place :-)

**Comment by hiker** [comment]
I'm just wondering: if I add a ``get_python`` script, would it be worth to install this in ``PSYCLONEHOME/bin`` (which would mean we could also use this in other scripts or Makefiles), or should I only use it for ``process.py``? I am just not sure if this is useful to make it available, i.e. if we actually have another script that might need it??  If so, then maybe we need a more unique name ... psyclone_get_python, since it would be in the user's path.

Unless I hear otherwise I will only implement this for ``process.py``, imho if required we can still move it later - but let me know if we already have a use case for it.

**Comment by hiker** [comment]
> python-is-python3

Ha, I just found out why my python suddenly became python2: I must have installed something that added ``python-is-python2``, after I installed ``python-is-python3``. So now I have both packages installed, and I guess whenever one of them is updated, my default python version changes. What a mess.

**Comment by arporter** [comment]
What about if we install the `process.py` script (with a suitable name change) like we do with `psyclone`. That way presumably the same magic re-writing of the shebang will be handled for us?

**Comment by hiker** [comment]
I can try, that would be easy. What about: ``process_jinja_psyclone``? That sounds pretty bad, but avoids the problem to clash with jinja or psyclone when using filename completion. 
We should probably also think about how to actually distribute PSydata libraries if a user only install files. The easy suggestion: we just copy the required files (as far as I can see the whole lib directory) 'somewhere', but document the fact that the user has to compile these files (F90= F90FLAGS= make all). Not sure what to do about dependencies (drhook etc). It would need a long list of environment variables for drhook, nvidia, .. ... not to mention, it might need infrastructure files (dl_esm_inf, LFRic). That would need to be specified as well *sigh*

**Comment by arporter** [comment]
Do you mean in order to support the case where people just do a `pip install psyclone`? The obvious answer is YAR (Yet Another  Repo) that could then also be a sub-module of PSyclone. Can anyone think of something better?

**Comment by hiker** [comment]
While this ticket should have been closed (sorry, my mistake), I actually need to reopen it. I have two machines on which it does not work as expected: 
1. One machine use a wrapper in ``(/usr)/bin/python`` (both locations!) that informs the user to please use python2 or python3. Since get_python first uses ``which python``, it starts the wrapper. Potential solution for this: test for python3 and python2 first, before python.
2. One other machine doesn't have ``realpath`` :( We can probably use ``readlink -f`` instead??

**Comment by hiker** [comment]
And one additional problem: when installing Intel's OpenAPI, it adds its own python 3.7 to the front of PATH. This resulted in psyclone tests/examples not working anymore (``get_python`` finds their python 3.7, but e.g. jinja is installed for python3.8, so 3.7 cannot find jinja).

So I adjust my suggestion above: ``get_python`` should search the paths in the following order:

1.  /usr/bin/python3
2. /usr/bin/python2
3. /usr/bin/python
4. which python3
5. which python2
6. which python

**Comment by hiker** [comment]
According to https://unix.stackexchange.com/questions/136494/whats-the-difference-between-realpath-and-readlink-f"
> On Linux systems, at the moment, your best bet to canonicalize a path that may contain symbolic links is readlink -f.

and

> readlink is installed within most Linux distributions. Whereas, realpath must often be explicitly installed.

So I'd suggest to also replace this accordingly.

**Comment by hiker** [comment]
OK, I tweaked the order to use /usr/bin/python{3,2} first - this now works on all machines that I can test, but it has the disadvantage that if a user should be using a python loaded from (say) a module environment (which we typically do at BOM), this loaded python would not be found :(  But the system python that is available works (you just might have to install additional packages like jinja using ``pip install --user`` ... which isn't that ideal either tbh). As mentioned before, if I use ``which python{3,2,}`` first (which would find a python installed from modules), it fails in case that you have Intel's OneAPI installed (or anything else that installs its own version of python at the front of the path).
This is implemented and would be ready to become a PR (branch 1001_part2_only_python3)

Only solution that I could think of at this stage that would be more general:
we support an environment variable, say ``$PSYCLONE_PYTHON``, which points to the python to use. So if the default python is not what you want, you could use this environment variable. Or is this an overkill? @arporter - feedback appreciated :grin:

**Comment by hiker** [comment]
Hmm - after a bit of thinking, there is one alternative: get_python could try to import some modules (jinja is the one I can think of), and only return the python version that successfully does the import.
That would slow this script down a bit, though hopefully not too bad. I tried a various systems. Typically under 0.1 seconds. Here from three systems:
```
 time python3 -c "import jinja2"

real	0m0.181s
user	0m0.052s
sys	0m0.073s

time python3 -c "import jinja2"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'jinja2'

real	0m0.045s
user	0m0.027s
sys	0m0.019s
time /usr/bin/python3 -c "import jinja2"

real	0m0.085s
user	0m0.084s
sys	0m0.001s
```
 Though our Cray system took 4 seconds (first time, then 0.35 seconds). I can't reproduce that now though, maybe the FS was just really slow (known issue with our module environment now and again).

**Comment by arporter** [comment]
I can see that looking at `/usr/bin` first will also cause problems when using Python virtual environments. Even checking whether the `jinja` module can be imported isn't totally fool proof if different versions of jinja are installed under different pythons.
I'm struggling to see the wood for the trees here - is this primarily for the 'build' system for the wrapper libraries? In which case, we just need to find a Python that has Jinja installed and we're done? If that's true then I think we have the search list as you suggest and then use the first one that has Jinja.
I'd be interested in what @TeranIvy thinks as she has the fun job of dealing with the MO environment...

**Comment by hiker** [comment]
So you would prefer to use ``which python{3,2,}`` first?

Yes, Iva's feedback would be great. It basically affects any non-standard packages that are needed and are installed by a user. It could even be fparser. Now for me I applied a simple work around - I installed e.g. jinja, sphinx etc (for my private account) several time in ``$HOME/.local``. So it doesn't matter if my python is now the one from intel (then it will look at the 3.7 subdirectory), or my system one (3.8). Plus - I now added intel to my standard search path in ``.bashrc``, so I also switched transparently to use Intel's python everywhere, which works fine after I had installed sphinx etc again.

I am happy to revisit the various platforms to see how difficult it is to make ``which ...`` first to work (e.g. on the Crays the system python does not even have pip iirc, so that makes it a bit hard to install something in ``.local`` - I will need to check if e.g. /usr/bin/python ``/some/other/path/pip`` would work - basically misusing a different pip binary). It might have been blown out of proportion a bit because by installing python I changed my default python.

**Comment by arporter** [comment]
Yes, I think I'd prefer to start with the `which` options as that seems most intuitive. We then go with the first Python we find that has jinja installed...?

**Comment by LonelyCat124** [comment]
Now we've removed python 2 support (and in general so have most OSes) do we still have this issue? @sergisiso @hiker @arporter
